### PR TITLE
add bulk forward geocoding

### PR DESCRIPTION
**Pull request by jeremyberglund-airspace**

### CODE REVIEW

**Review by jclinford — APPROVED** [review]

File: geocoding.go [hunk 1]
```diff
@@ -27,6 +27,12 @@ type ReverseGeocodeRequest struct {
 	Types    Types  `json:"types,omitempty"`
 }
 
+type ForwardGeocodeBatchRequest []ForwardGeocodeRequest
```
Comment (ravisastryk) [inline geocoding.go:30]: Consider renaming `ForwardGeocodeBatchRequest` to something to indicate of its bulk processing nature for clarity. How about `BulkForwardGeocodeRequest`?

File: geocoding.go [hunk 2]
```diff
@@ -40,6 +46,38 @@ type ForwardGeocodeRequest struct {
 	Types        Types
 }
 
+func (r ForwardGeocodeRequest) MarshalJSON() ([]byte, error) {
+	type forwardGeocodeRequest struct {
+		SearchText   string       `json:"q,omitempty"`
```
Comment (ravisastryk) [inline geocoding.go:51]: Check for unintended single-letter variables, as they may reduce code readability. `q` can also be interpreted in some cases for query. Would you like to  be `searchtext` or `q`? Please check again and confirm.

File: geocoding.go [hunk 3]
```diff
@@ -40,6 +46,38 @@ type ForwardGeocodeRequest struct {
 	Types        Types
 }
 
+func (r ForwardGeocodeRequest) MarshalJSON() ([]byte, error) {
+	type forwardGeocodeRequest struct {
```
Comment (ravisastryk) [inline geocoding.go:50]: nit: It's generally a good practice to group similar fields(eg: strings together, bool together, slices together: "like birds of same feather flock together" :-)  for better readability.

File: geocoding.go [hunk 4]
```diff
@@ -169,6 +207,29 @@ func forwardGeocode(ctx context.Context, client *Client, req *ForwardGeocodeRequ
 	return &response, nil
 }
 
+// https://docs.mapbox.com/api/search/geocoding/#batch-geocoding
+func forwardGeocodeBatch(ctx context.Context, client *Client, req ForwardGeocodeBatchRequest) (*GeocodeBatchResponse, error) {
+	query := url.Values{}
```
Comment (ravisastryk) [inline geocoding.go:212]: `url.Values` setup in f`orwardGeocode`, `reverseGeocode`, and batch geocoding functions has redundancy.

May be introduce a helper function something like this one below and use it here?
```
func prepareForwardQuery(req ForwardGeocodeRequest, apiKey string) url.Values {
    query := url.Values{}
    query.Set("q", req.SearchText)
    query.Set("access_token", apiKey)
    query.Set("autocomplete", strconv.FormatBool(req.Autocomplete))

    if !req.BBox.Min.IsZero() {
        query.Set("bbox", req.BBox.query())
    }
    if req.Country != "" {
        query.Set("country", req.Country)
    }
    if req.Language != "" {
        query.Set("language", req.Language)
    }
    if req.Limit != 0 {
        query.Set("limit", strconv.Itoa(req.Limit))
    }
    if !req.Proximity.IsZero() {
        query.Set("proximity", req.Proximity.WGS84Format())
    }
    if len(req.Types) != 0 {
        query.Set("types", req.Types.query())
    }

    return query
}

func forwardGeocode(ctx context.Context, client *Client, req *ForwardGeocodeRequest) (*GeocodeResponse, error) {
    query := prepareForwardQuery(*req, client.apiKey)

    apiResponse, err := client.get(ctx, GeocodingForwardEndpoint, query)
    if err != nil {
        return nil, fmt.Errorf("failed to get forward geocode response: %w", err)
    }

    var response GeocodeResponse
    if err := client.handleResponse(apiResponse, &response, GeocodingRateLimit); err != nil {
        return nil, fmt.Errorf("failed to handle forward geocode response: %w", err)
    }

    return &response, nil
}
```

File: geocoding.go [hunk 5]
```diff
@@ -169,6 +207,29 @@ func forwardGeocode(ctx context.Context, client *Client, req *ForwardGeocodeRequ
...
+	query.Set("access_token", client.apiKey)
+
+	b, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
```
Comment (ravisastryk) [inline geocoding.go:217]: How about providing better error handling and make it descriptive?
```
        return nil, fmt.Errorf("error marshalling forward geocode batch request: %w", err)

```

File: geocoding.go [hunk 6]
```diff
@@ -169,6 +207,29 @@ func forwardGeocode(ctx context.Context, client *Client, req *ForwardGeocodeRequ
...
+	}
+
+	apiResponse, err := client.post(ctx, GeocodingBatchEndpoint, query, bytes.NewReader(b))
+	if err != nil {
+		return nil, err
```
Comment (ravisastryk) [inline geocoding.go:222]: How about informative error?
```
        return nil, fmt.Errorf("post request to geocoding batch endpoint failed: %w", err)

```

File: geocoding.go [hunk 7]
```diff
@@ -169,6 +207,29 @@ func forwardGeocode(ctx context.Context, client *Client, req *ForwardGeocodeRequ
...
+	}
+
+	var response *GeocodeBatchResponse
+	if err := client.handleResponse(apiResponse, &response, GeocodingRateLimit); err != nil {
+		return nil, err
```
Comment (ravisastryk) [inline geocoding.go:227]: Here too
```
        return nil, fmt.Errorf("error handling response for forward geocode batch request: %w", err)

```

**Pull request merged**